docs/namespace_config_spec.md

```diff
@@ -0,0 +1,319 @@
+  some(iteratorfn: (element: Namespace) => boolean): boolean;
+}
+```
+
+### Operators
+
+The following character sequences represent boolean operators:
+
+| Operator | Signature | Semantic |
+| --- | --- | --- |
+| `&&` | *x* `&&` *y* | true iff. *x* and *y* are true |
+| `\|\|` | *x* `\|\|` *y* | true iff. *x* or *y* are true |
+
+The following character sequences represent miscellaneous operators:
+
+| Operator | Example | Semantic |
+| --- | --- | --- |
+| `(`, `)` | `x()` | call function `x` |
+| `[`, `]` | `T[]` | `T` is an array type |
+| `{`, `}` | `class x {}` | Scope delimiters |
+| `=>` | `list.some(el => f(el))` | Lambda definition token. |
+| `.` | `this.x` | Property traversal token |
+| `:` | `relation: type` | Relation/type separator token |
+| `=` | `permits = {...}` | Assignment token |
+| `,` | `{x: 1, y: 2}` | Property separator |
+| `'` | `'string'` | Single quoted string literal |
+| `"` | `"string"` | Double quoted string literal |
+
+## Statements
+
+### Type declaration
+
+The top level of the configuration consists of a list of `class` declaration for
+each namespace. Each `class` consists of relation declarations and permission
+declarations.
+
+```ebnf
+Config          = [ ClassDecl ] .
+ClassDecl       = "class" identifier "implements" "Namespace" "{" ClassSpec "}" .
+ClassSpec       = MetadataDefn | [ RelationDecls ] | [ PermissionDefns] .
+```
+
+#### Example
+
+The following declares a type *User* that is never the object in any
+relation. It can, however, be the subject of a relation, the relation being
+declared on some other type.
+
+```ts
+class User implements Namespace { 
+  metadata = { 
+    id: "1"
+  }
+}
+```
+
+
+### Metadata definiton
+
+The namespace metadata is defined through an object property of the class.
+Currently, there is only the `id` metadata field, which specifies the unique identifier for the namespace.
+This way, namespaces can be renamed without needing to update the underlying relation tuples.
+
+```ebnf
+MetadataDefn   = "metadata" "=" "{" id ":" string_lit "}" .
+```
+
+### Relation declaration
+
+Inside the class declaration, relations are declared. `RelationName` must be
+unique identifiers and are used as the relation strings in the tuples. The
+`TypeName` must be the name of a `type` that is defined above or below (in
+TypeScript: a class that implements `Namespace`).
+
+```ebnf
+RelationDecls   = "related" "=" "{" { RelationName ":" TypeName "[]" } "}" .
+RelationName    = literal .
+TypeName        = literal .
+```
+
+Note that all relations are defined as array types `T[]`, because there are
+naturally only many-to-many relations in Keto.
+
+The following declares a type *Document* with three relations: *owners* and
+*viewers*, both of which have *users* as subjects. Additionally, the relation
+*parent* has type *Document*.
+```ts
+class User {
+  metadata = {
+    id: "1"
+  }
+}
+
+class Document {
+  metadata = {
+    id: "2"
+  }
+  related = {
+    parent: Document[]
```

Comment:
Yes, this is more correct. We can't specify that we want only one, or exactly one parent in Keto, since there are always just relation tuples in the database (at least these constraints would involve fundamental changes to the data store).

We should encourage users to always pluralize their relations, as already suggested by the now-mandatory array type.